### PR TITLE
Add typing for all of myhoard and enable check_untyped_defs

### PR DESCRIPTION
**Pull request by jelmer**
I've added some asserts to make mypy happy; these all check internal constraints, rather than external expectations. Happy to add more extended descriptions to them, if deemed useful - but I think most of them are fairly clear if we ever trigger them.

TypedDict doesn't have a derived type for its keys yet and I didn't want to add a static Literal for each TypedDict that we have to keep up to date, so there's some additional "type: ignore"s for now. See also https://github.com/python/mypy/issues/11553

### CODE REVIEW

File: myhoard/controller.py [hunk 1]
```diff
@@ -1152,7 +1178,11 @@ def _mark_periodic_backup_requested_if_interval_exceeded(self):
         half_backup_interval_s = self.backup_settings["backup_interval_minutes"] * 60 / 2
         if (
             last_normalized_backup_time != normalized_backup_time
-            and self.state["backup_request"].get("normalized_backup_time") != normalized_backup_time
+            and (
+                not self.state["backup_request"]
+                or self.state["backup_request"]["normalized_backup_time"]  # pylint: disable=unsubscriptable-object
```
Comment (kmichel-aiven) [inline myhoard/controller.py:1183]: Here too, why mypy can't see that "backup_request" has the type we expect ?

**Review by kmichel-aiven — DISMISSED** [review]
It would help the review to split this in multiple commits.

Not one commit per tiny change (:zzz:), but at least separate the mass of trivial "add type hint" and the more difficult cases where the code had to be changed in addition to the hints, or if there was some tradeoff involved.

**Comment by jelmer** [comment]
> It would help the review to split this in multiple commits.
> 
> Not one commit per tiny change (zzz), but at least separate the mass of trivial "add type hint" and the more difficult cases where the code had to be changed in addition to the hints, or if there was some tradeoff involved.

Split out the changes for some of the building blocks:

* https://github.com/aiven/myhoard/pull/125 (util)
* https://github.com/aiven/myhoard/pull/126 (basebackup)
* https://github.com/aiven/myhoard/pull/127 (statemanager)

**Comment by jelmer** [comment]
> > It would help the review to split this in multiple commits.
> > Not one commit per tiny change (zzz), but at least separate the mass of trivial "add type hint" and the more difficult cases where the code had to be changed in addition to the hints, or if there was some tradeoff involved.
> 
> Split out the changes for some of the building blocks:
> 
> * [util: Add more typing #125](https://github.com/aiven/myhoard/pull/125) (util)
> * [basebackup: Add more typing #126](https://github.com/aiven/myhoard/pull/126) (basebackup)
> * [statemanager: add typing #127](https://github.com/aiven/myhoard/pull/127) (statemanager)

And further:

* https://github.com/aiven/myhoard/pull/128 (backup_stream)
* https://github.com/aiven/myhoard/pull/129 (restore_coordinator)

Once those have landed, I suspect this PR (which I'll update) will be a bit more managable.

File: myhoard/backup_stream.py [hunk 2]
```diff
@@ -1023,30 +1087,30 @@ def _upload_binlog(self, binlog):
...
-        binlog = dict(binlog)
-        binlog["remote_index"] = next_index
-        binlog["remote_key"] = index_name
-        binlog["compression_algorithm"] = compression_algorithm
-        metadata = make_fs_metadata(binlog)
-        self.current_upload_index = binlog["local_index"]
+        remote_binlog: RemoteBinlogInfo = cast(
+            RemoteBinlogInfo,
+            dict(binlog, remote_index=next_index, remote_key=index_name, compression_algorithm=compression_algorithm),
+        )
```
[the remark this answers is not in the document]
Comment (jelmer) [reply myhoard/backup_stream.py:1100]: Unfortunately it doesn't appear so - I've added a comment.

File: myhoard/controller.py [hunk 3]
```diff
@@ -234,21 +243,20 @@ def is_safe_to_reload(self) -> bool:
...
-        }
+        backup_time: str = normalized_backup_time or self._current_normalized_backup_timestamp()
+        new_request: BackupRequest = {"backup_reason": backup_reason, "normalized_backup_time": backup_time}
         with self.lock:
             if self.state["backup_request"]:
                 old_request: BackupRequest = self.state["backup_request"]
                 if (
                     old_request == new_request
-                    or new_request["normalized_backup_time"] < old_request["normalized_backup_time"]
+                    or backup_time < old_request["normalized_backup_time"]  # pylint: disable=unsubscriptable-object
```
[the remark this answers is not in the document]
Comment (jelmer) [reply myhoard/controller.py:253]: It appears to be a bug in pylint. There are a few similar bugs in its bug tracker, but nothing that isn't fixed and matches this specific issue. mypy is fine with it.

I'll try to spend some time to try and reproduce it with a minimal test case & file an upstream bug. Ideally that shouldn't prevent this entire branch from landing though.

File: myhoard/controller.py [hunk 4]
```diff
@@ -1200,7 +1230,7 @@ def _purge_old_backups(self):
         # For simplicity only ever drop one backup here. This function
         # is called repeatedly so if there are for any reason more backups
         # to drop they will be dropped soon enough
-        purgeable = sorted(purgeable, key=lambda info: info["completed_at"])
+        purgeable = sorted(purgeable, key=lambda info: cast(float, info["completed_at"]))
```
[the remark this answers is not in the document]
Comment (jelmer) [reply myhoard/controller.py]: Resolved by factoring out this (common) pattern into a function.

File: myhoard/controller.py (hunk 1 above)
Comment (jelmer) [reply myhoard/controller.py:1180]: mypy has no problems with it, but pylint does for some reason - same as above.

**Review by kmichel-aiven — APPROVED** [review]

**Pull request merged**